GH-41664: [C++][Python] PrettyPrint non-cpu data by copying to default CPU device

jorisvandenbossche opened the pull request:
### Rationale for this change

The various Python reprs or the C++ `PrettyPrint` functions will currently just segfault when passing an object that has its data on a non-CPU device. In python, getting a segfault while displaying the object is very annoying, and so we should make this at least not crash.

### What changes are included in this PR?

When we detect data on a non-CPU device passed to `PrettyPrint`, we copy the necessary part (the full Arrays for Array/RecordBatch, or the full chunks that are being printed for ChunkedArray/Table) to the default CPU device, and then use the existing print utilities as is on this copied subset.

For large data, this can be potentially costly by copying a lot of data (but you can always avoid that by not printing the data), but for chunked data we will still only copy those chunks of the full dataset needed to print the object. 
Longer term, we should investigate if we can actually copy sliced arrays to a different device (with actual pruning of the buffers while copying): https://github.com/apache/arrow/issues/43055

### Are these changes tested?

Yes

### Are there any user-facing changes?

No
* GitHub Issue: #41664

jorisvandenbossche commented:
@pitrou I started looking into this on the C++ side, but with the current approach of slice+copy+concat, I could also easily do this just on the Python side before passing the data to `PrettyPrint`. But do you think this is useful to keep on the C++ side?

pitrou commented:
Yes, it's certainly useful on the CPU side as well.

jorisvandenbossche commented on python/pyarrow/tests/test_cuda.py:
```diff
@@ -792,3 +792,57 @@ def test_IPC(size):
+    assert str(carr) == str(arr)
+
+
+def make_chunked_array(n_elements_per_chunk, n_chunks):
+    arrs = []
+    carrs = []
+    for _ in range(n_chunks):
+        batch = make_recordbatch(n_elements_per_chunk)
+        cbuf = cuda.serialize_record_batch(batch, global_context)
+        cbatch = cuda.read_record_batch(cbuf, batch.schema)
+        arrs.append(batch["f0"])
+        carrs.append(cbatch["f0"])
+
+    return pa.chunked_array(arrs), pa.chunked_array(carrs)
```
This is still a bit verbose, but can be simplified once the "copy to device" functionality is exposed (should maybe do that first now)

jorisvandenbossche commented:
I still need to add C++ tests, but this is all working from testing it with pyarrow.

jorisvandenbossche commented:
@github-actions crossbow submit test-cuda-python

jorisvandenbossche commented on cpp/src/arrow/pretty_print.cc:
```diff
@@ -413,18 +414,36 @@ Status ArrayPrinter::WriteValidityBitmap(const Array& array) {
   }
 }
 
+Result<std::shared_ptr<Array>> CopyStartEndToCPU(const Array& arr, int window) {
+  std::shared_ptr<Array> arr_sliced;
+  if (arr.length() > (2 * window + 1)) {
+    ARROW_ASSIGN_OR_RAISE(auto arr_start,
+                          arr.Slice(0, window + 1)->CopyTo(default_cpu_memory_manager()));
+    ARROW_ASSIGN_OR_RAISE(
+        auto arr_end,
+        arr.Slice(arr.length() - window - 1)->CopyTo(default_cpu_memory_manager()));
+    ARROW_ASSIGN_OR_RAISE(arr_sliced, Concatenate({arr_start, arr_end}));
+  } else {
+    ARROW_ASSIGN_OR_RAISE(arr_sliced, arr.CopyTo(default_cpu_memory_manager()));
```
@pitrou I am using the slice+copy approach above in the idea of only copying a small part of the full array to the CPU device. However, on second though, I assume that this actually doesn't work as intended? The `Slice` is zero-copy (just adding offsets), and I assume the `CopyTo` just naively copies all buffers of the array in its entirety, not actually truncating and copying only the part of the buffer that is needed?

pitrou replied:
Hmm, perhaps @felipecrv can answer these questions.

felipecrv replied:
> CopyTo just naively copies all buffers of the array in its entirety, not actually truncating and copying only the part of the buffer that is needed?

Yes, because `MemoryManager`s work in terms of Buffers, not ranges. We recently added a function that can copy slices to CPU, but it falls back to full buffer copy if the source `MemoryManager` is not a CPU memory manager.

The protocols for memory movement between devices might not allow for range-based copies (@zeroshade might know better than me), so instead expanding the `MemoryManager` interface we should probably have a way to slice the array where it is (device-specific) and then copy the smaller array. (?)

felipecrv replied:
Writing a slicer in CUDA might be a ton of work though.

zeroshade replied:
@felipecrv why does it fallback to a full buffer copy if the source `MemoryManager` isn't a CPU memory manager? Couldn't it just use `cuMemCpy` with the appropriate offset into the buffer?

zeroshade replied:
Oh, you're referring to slicing the *array* without copying the entirety of the buffers. Yea, that's a trickier proposition. The functions exist in libcudf already to do this, but we don't link against libcudf in libarrow_cuda (nor do I think we should), but in theory we *could* write the necessary code or at least figure it out by looking at how libcudf does it. For now, I think that the naive approach is likely fine until someone actively states they have an issue

jorisvandenbossche replied:
> For now, I think that the naive approach is likely fine until someone actively states they have an issue

So for the repr, there are basically two options:

- Don't print any actual data (eg in the Array repr, instead of including the data, add a general message like "xx values on xx device")
- Just do what I am doing here, acknowledging that the repr therefore _can_ be costly (if this is a problem, the user can still ensure to not actually use this / print the object). In practice, very often you will have chunked data, and so for example when having a table with many smaller chunks, we will still only copy a couple of chunks.

@zeroshade I take your comment as being fine with the second option?

zeroshade replied:
I'm fine with either approach. As long as we're able to indicate the situation without crashing, we're good. Acknowledging that the repr *can* be costly is a good tradeoff

jorisvandenbossche commented:
@github-actions crossbow submit test-cuda-python

felipecrv commented on cpp/src/arrow/pretty_print.cc:
```diff
@@ -476,8 +479,7 @@ Status PrettyPrint(const ChunkedArray& chunked_arr, const PrettyPrintOptions& op
       chunk_options.indent += options.indent_size;
-      ArrayPrinter printer(chunk_options, sink);
-      RETURN_NOT_OK(printer.Print(*chunked_arr.chunk(i)));
+      RETURN_NOT_OK(PrettyPrint(*chunked_arr.chunk(i), chunk_options, sink));
```
This is creating a new `ArrayPrinter` on every iteration. A cleaner fix here would be changing `Status Print(const Array& array) { ... }` right before `RETURN_NOT_OK(VisitArrayInline(array, this));`

Constructing an `ArrayPrinter` is very cheap at the moment, but that might not forever be the case.

jorisvandenbossche commented:
@github-actions crossbow submit test-cuda-python

pitrou commented:
> we copy the necessary part (window size of start and end of array) to the default CPU device

Where does this happen? Judging by the PR diff, it seems it is copying the entire contents, doesn't it?

jorisvandenbossche commented:
> Judging by the PR diff, it seems it is copying the entire contents, doesn't it?

Sorry, I should have updated the top description of the PR. I initially thought I was doing that (copying only the necessary parts), but that's not actually what copying a sliced array does, see the discussion in https://github.com/apache/arrow/pull/42010#discussion_r1638172658 
(will update the top post now)

pitrou commented:
I mean that the array isn't sliced at the point where you're copying it, is it?

jorisvandenbossche commented:
I was doing that in the initial version (that the now updated top post described), but since slicing doesn't reduce the amount of data being copied, I removed that (https://github.com/apache/arrow/pull/42010/commits/595ad8982ae19d09a7ee0d2eabbd1dbce28ca7c8) (again see the discussion at https://github.com/apache/arrow/pull/42010#discussion_r1638172658)

pitrou commented:
Ahah, I see. We should open followup issues then, if not already done!

felipecrv reviewed (approved):
LGTM

jorisvandenbossche commented:
@github-actions crossbow submit test-cuda-python

jorisvandenbossche commented:
Opened https://github.com/apache/arrow/issues/43055 as a follow-up and linked that from the top post (and will add a TODO mentioning that issue in a comment)

felipecrv commented on cpp/src/arrow/pretty_print.cc:
```diff
@@ -391,7 +391,14 @@ class ArrayPrinter : public PrettyPrinter {
   Status Print(const Array& array) {
-    RETURN_NOT_OK(VisitArrayInline(array, this));
+    if (array.device_type() != DeviceAllocationType::kCPU) {
+      // TODO: ideally we only copy start/end slice based on the window size that is being
+      // printed (requires CopyTo for slices https://github.com/apache/arrow/issues/43055)
```
```suggestion
      // GH-43055: ideally we only copy start/end slices from non-CPU memory
      // based on the window size that is being printed
```
